Fix not terminating TaskInstances when `enable_logging=True` in DockerSwarmOperator

stavdav143 opened the pull request:
closes: #28452

Changing the way docker service logs are consumed by using a buffer and repeatedly querying the docker API without the 'follow' option. This 'follow' option  makes the request hang waiting for input even if the service has ended and can only be resolved upstream by the native docker API. We bypass this deficiency by polling the API.
Tested on a production DockerSwarm cluster. 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

stavdav143 committed:
Issue: TaskInstances would not succeed when using enable_logging=True option in DockerSwarmOperator. #28452
https://github.com/apache/airflow/issues/28452

Changing the way docker service logs are consumed by using a buffer and repeatedly querying the docker API without the 'follow' option. This 'follow' option  makes the request hang waiting for input even if the service has ended
and can only be resolved upstream by the native docker API.

hussein-awala commented on airflow/providers/docker/operators/docker_swarm.py:
```diff
@@ -178,31 +178,39 @@ def _has_service_terminated(self) -> bool:
+
+# A custom logs bytes buffer that keeps the current state of the
+# docker service logs, as these get continuously polled by the client.
+class LogsBuffer:
```
can you make a docstring from these comments and delete them?

hussein-awala commented on airflow/providers/docker/operators/docker_swarm.py:
```diff
@@ -178,31 +178,39 @@ def _has_service_terminated(self) -> bool:
+# docker service logs, as these get continuously polled by the client.
+class LogsBuffer:
+    def __init__(self) -> None:
+        self.cur_buffer = bytes()
```
You can store the last printed index instead of storing the buffer in memory, WDYT?

stavdav143 replied:
Good point @hussein-awala , do you want to take a look now? Again, only been tested on a production swarm without the airflow unit tests.

stavdav143 committed:
Merge branch 'apache:main' into main

stavdav143 committed:
Issue: TaskInstances would not succeed when using enable_logging=True option in DockerSwarmOperator. #28452
    https://github.com/apache/airflow/issues/28452
    Changing the way docker service logs are consumed by using a buffer and repeatedly querying the docker API without the 'follow' option. This 'follow' option  makes the request hang waiting for input even if the service has ended
    and can only be resolved upstream by the native docker API.
Update code based on hussein-awala's input.

stavdav143 committed:
Merge branch 'main' into main
